### PR TITLE
Add destroy server group ops and other minor bug fixes

### PR DESCRIPTION
**Pull request by milismsft**
Add destroy server group ops. 
Cleanup resources that are created as part of the create server group in case of a failure during the creation.
Add retries support when deleting a resource in Azure (work around for random timeouts when accessing Azure resources in the Azure Cloud).
Fix an exception in clouddriver console when the application has no clusters.
Add type as part of the deployment name.

### CODE REVIEW

**Comment by milismsft** [comment]
@spinnaker/netflix-reviewers PTAL

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy [hunk 1]
```diff
@@ -154,6 +155,38 @@ public class AzureComputeClient extends AzureBaseClient {
     null
   }
 
+  ServiceResponse<Void> destroyServerGroup(String resourceGroupName, String serverGroupName) {
+    ServiceResponse<Void> result = null
+    // The API call might return a timeout exception or some other Azure CloudException that is not
+    //   related to the operation we are trying to execute; retry couple of times and if the final
+    //   retry fails then rethrow
+    long operationRetry = 0
+    while (operationRetry < AZURE_ATOMICOPERATION_RETRY) {
```
Comment (ajordens) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy]: You probably want to introduce a sleep between cycles, and perhaps some logging to know that retries are happening.

See `AmazonClientInvocationHandler` for an example of how we've done this similarly.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy [hunk 2]
```diff
@@ -184,13 +184,83 @@ class AzureNetworkClient extends AzureBaseClient {
...
     def publicIpAddressName = AzureUtilities.getResourceNameFromID(loadBalancer.frontendIPConfigurations.first().getPublicIPAddress().id)
-    client.getLoadBalancersOperations().delete(resourceGroupName, loadBalancerName)
 
-    client.getPublicIPAddressesOperations().delete(resourceGroupName, publicIpAddressName)
+    ServiceResponse<Void> result = null
+    // The API call might return a timeout exception or some other Azure CloudException that is not
+    //   related to the operation we are trying to execute; retry couple of times and if the final
+    //   retry fails then rethrow
+    long operationRetry = 0
+    while (operationRetry < AZURE_ATOMICOPERATION_RETRY) {
```
Comment (ajordens) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: Looks like this behavior is used in a handful of places, I'd suggest a method (taking a closure) that wraps the retry behavior and just calls out to the closure to actually execute the client call.

This is primarily towards eliminating copy & pasted code.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy [hunk 3]
```diff
@@ -62,12 +62,13 @@ class AzureResourceManagerClient extends AzureBaseClient {
...
                                                 String region,
                                                 String resourceName,
+                                                String resourceType,
                                                 Map<String, String> templateParams = [:]) {
 
     // TODO validate that all callers invoke this themselves, then remove this call
     initializeResourceGroupAndVNet(credentials, resourceGroupName, null, region)
 
-    String deploymentName = resourceName + AzureUtilities.NAME_SEPARATOR +"deployment"
+    String deploymentName = resourceName + AzureUtilities.NAME_SEPARATOR + resourceType + AzureUtilities.NAME_SEPARATOR +"deployment"
```
Comment (ajordens) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy]: Not a big deal but `[resourceName, resourceType, "deployment"].join(AzureUtilities.NAME_SEPARATOR]` would be another way to write this.

**Comment by ajordens** [comment]
Few comments but this is all within azure packages so LGTM.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy (hunk 1 above)
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy]: Makes sense; I can add the sleeps in between the calls.
BTW can you point me into AmazonClientInvocationHandler what method should I check? I did not notice anyone that has similar retries.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy (hunk 2 above)
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureNetworkClient.groovy]: I've made a note to refactor these into one single method; I've initially started with something much simpler than added the extra lines.

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy (hunk 3 above)
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureResourceManagerClient.groovy]: Makes sense; I'll do the change.

**Comment by milismsft** [comment]
@spinnaker/netflix-reviewers @ajordens  PTAL

File: clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy (hunk 1 above)
Comment (milismsft) [inline clouddriver-azure/src/main/groovy/com/netflix/spinnaker/clouddriver/azure/client/AzureComputeClient.groovy]: @ajordens I moved most of the duplicated code that handles the retries into the base class; the two methods will retry on both getting the operation object and the actual delete op. 
In time I will continue to build on top of that and change some more operations to use the same pattern (see get like methods).

**Comment by ajordens** [comment]
Needs a quick little rebase

**Comment by milismsft** [comment]
@ajordens  rebase done; can you help with merging it in?

**Pull request merged**